File: Battleships/java/src/main/java/akuna/battleships/Connection.java
package akuna.battleships;

import java.io.*;
import java.net.Socket;

class Connection implements AutoCloseable {
    private static int DEFAULT_PORT = 50008;

    private final Socket socket;
    private final BufferedReader reader;
    private final Writer writer;

    private Connection(String host, int port) throws IOException {
        System.out.println(String.format("Connecting to %s:%d", host, port));
        this.socket = new Socket(host, port);
        this.reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        this.writer = new OutputStreamWriter(this.socket.getOutputStream());
    }

    Connection(String host) throws IOException {
        this(host, DEFAULT_PORT);
    }

    void send(String msg) throws IOException {
        System.out.println(String.format("> %s", msg));
        this.writer.write(msg);
        this.writer.write(System.lineSeparator());
        this.writer.flush();
    }

    String getLine() throws IOException {
        String line;
        do {
            line = this.reader.readLine();
            System.out.println(String.format("< %s", line));
            if (line == null) {
                return null;
            }
        } while (line.isEmpty() || line.charAt(0) == ' ');
        return line;
    }

    public void close() throws IOException {
        this.writer.close();
        try {
            this.reader.close();
        } finally {
            this.socket.close();
        }
    }
}

File: Battleships/csharp/src/Akuna/Battleships/Connection.cs
using System;
using System.IO;
using System.Net.Sockets;

namespace Akuna.Battleships
{
    public class Connection : IDisposable
    {
        private static readonly int DEFAULT_PORT = 50008;

        private readonly TcpClient _client;
        private readonly StreamReader _reader;
        private readonly StreamWriter _writer;

        public Connection(String host, int port)
        {
            System.Console.WriteLine($"Connecting to {host}:{port}");
            _client = new TcpClient(host, port);
            _reader = new StreamReader(_client.GetStream());
            _writer = new StreamWriter(_client.GetStream());
        }

        public Connection(String host)
            : this(host, DEFAULT_PORT)
            { }

        public void Send(String msg)
        {
            Console.WriteLine($"> {msg}");
            _writer.WriteLine(msg);
            _writer.Flush();
        }

        public String GetLine()
        {
            String line;
            do
            {
                line = _reader.ReadLine();
                Console.WriteLine($"< {line}");
                if (line == null)
                {
                    return null;
                }
            }
            while (line == "" || line[0] == ' ');
            return line;
        }

        public void Dispose()
        {
            _writer.Close();
            _reader.Close();
            _client.Close();
        }
    }
}
